src/lsp/cobol_parser/expect/gen_expect.ml

```diff
@@ -0,0 +1,421 @@
+    match rhs.(idx) with
+      | N nt, _, _ -> List.filter_map terminal_filter_map (Nonterminal.first nt)
+      | T _, _, _ -> begin
+        match continue_while_terminal rhs idx [] with
+          | [] -> []
+          | l -> [K (NEL.of_list l)] end
+      | exception Invalid_argument _ -> []
+  in
+  Lr1.tabulate begin fun lr1 ->
+    let custom_comp_entries = CompEntrySet.from_list_map
+      ~add:CompEntrySet.add_option
+      begin function
+        | N nonterm, _ -> nonterminal_filter_map nonterm
+        | _ -> None end
+      (Lr1.transitions lr1) in
+    CompEntrySet.from_list_map
+      ~init:custom_comp_entries ~add:CompEntrySet.add_list
+      next_tokens_from_item (Lr0.items @@ Lr1.lr0 lr1)
+  end
+
+let emit_acceptable_terminals_in_env ppf = (* taking transitions *)
+  Fmt.pf ppf {|
+let acceptable_terminals_in ~env : completion_entry list =
+  NEL.(match Menhir.current_state_number env with
+|};
+  Lr1.fold (fun lr1 acc ->
+    match completion_entries_of lr1 with
+      | s when CompEntrySet.is_empty s -> acc
+      | s -> CompEntryMap.add_to_list s lr1 acc)
+  CompEntryMap.empty
+  |> inv (CompEntryMap.fold (fun s l acc -> (l, CompEntrySet.elements s)::acc)) []
+  |> pp_match_cases ppf
+    Fmt.(using Lr1.to_int int)
+    (pp_brackets @@ pp_list pp_completion_entry)
+    "[])"
+
+let guess_default_value ~mangled_name ~typ =
+  match typ, mangled_name with
+  | _ when EzString.ends_with ~suffix:"option" typ ||
+          EzString.starts_with ~prefix:"option_" mangled_name
+  -> Some "None"
+  | _ when EzString.ends_with ~suffix:"list" typ ||
+  EzString.starts_with ~prefix:"loption_" mangled_name ||
+  EzString.starts_with ~prefix:"list_" mangled_name
+  -> Some "[]"
+  | _ when EzString.ends_with ~suffix:"unit" typ
+  -> Some "()"
+  | _ when EzString.ends_with ~suffix:"bool" typ ||
+  EzString.starts_with ~prefix:"boption_" mangled_name
+  -> Some "false"
+  | _ -> None
+
+let is_nullable_without_recovery = Nonterminal.tabulate begin fun nt ->
+  Nonterminal.nullable nt &&
+    Nonterminal.attributes nt
+    |> List.find_opt (Attribute.has_label "recovery")
+    |> Option.is_none
+  end
+
+let best_guess_default_value = Nonterminal.tabulate begin fun nt ->
+  if not (is_nullable_without_recovery nt)
+  then None
+  else
+  let mangled_name = Nonterminal.mangled_name nt in
+  let guessed_default = Option.bind
+    (Nonterminal.typ nt)
+    (fun typ -> guess_default_value ~mangled_name ~typ) in
+  let given_default = Nonterminal.attributes nt
+    |> List.find_opt (Attribute.has_label "default")
+    |> Option.map Attribute.payload in
+  match given_default with
+    | Some _ -> given_default
+    | None -> guessed_default
+  end
+
+let emit_default_value_of_nullables ppf =
+  Fmt.pf ppf {|
+let guessed_default_value_of_nullables (type a): a Menhir.symbol -> a = function
+  (* If this function does not compile it probably means you're missing a
+     [@default] attribute on a nullable nonterminal in the grammar *)
```

Comment:
Does this comment still hold (there'e a wildcard with `raise Not_found`) ?

Reply:
I'm also wondering why this function takes a `symbol`, whereas it only makes sense on `nonterminal`.

Reply:
Yes it holds, I put it because on some cases the guessing the type part fails (ex: type is `int * int option` -> defaults to `None` instead of `(0, None)`). I updated the comment for clarity

It returns a symbol for consistency with `Recover.default_value` and to simplify the logic on the `lsp_completion` side of things

Reply:
I believe `symbol` is defines as `type symbol = T of terminal | N of nonterminal`. So passing `(N nt)` to `Recover.default_value` is not an issue. We can postpone that change of API for later if you prefer.